src/aten/sycl/DistributionTemplates.h

```diff
@@ -0,0 +1,595 @@
+    double from_,
+    double to_,
+    RNG gen) {
+  AT_DISPATCH_FLOATING_TYPES_AND2(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      iter.dtype(),
+      "uniform_kernel_xpu",
+      [&] {
+        auto from = static_cast<scalar_t>(from_);
+        auto to = static_cast<scalar_t>(to_);
+        using opmath_t = at::opmath_type<scalar_t>;
+        auto range = static_cast<opmath_t>(to - from);
+        uniform_and_transform<scalar_t, opmath_t, rand4_engine_calls>(
+            iter,
+            gen,
+            UniformTransformFunctor<scalar_t, opmath_t>(range, from, to));
+      });
+}
+
+// ====================== Bernoulli ======================
+
+template <typename scalar_t>
+struct BernoulliFunctor {
+  scalar_t operator()(scalar_t out, scalar_t p) const {
+    return static_cast<scalar_t>(out < p);
+  }
+};
+
+template <typename RNG>
+void bernoulli_kernel(const TensorBase& self, const TensorBase& p_, RNG gen) {
+  TORCH_CHECK(
+      at::isFloatingType(p_.scalar_type()),
+      "expected probabilities tensor to have floating type, got ",
+      p_.scalar_type());
+  const auto p_type = self.dtype() == at::kDouble ? at::kDouble : at::kFloat;
+  auto p_xpu = p_.to(TensorOptions().device(self.device()).dtype(p_type));
+  auto p = expand_inplace(self, p_xpu);
+
+  Tensor self_float;
+  auto self_type = self.scalar_type();
+  if (!(self_type == at::ScalarType::Float ||
+        self_type == at::ScalarType::Double))
+    self_float = at::empty(self.sizes(), self.options().dtype(at::kFloat));
+  else
+    self_float = self;
+
+  auto iter_uniform = at::TensorIterator::borrowing_nullary_op(self_float);
+  uniform_kernel<RNG>(iter_uniform, 0.0, 1.0, gen);
+
+  auto iter = TensorIteratorConfig()
+                  .add_output(self)
+                  .add_input(self_float)
+                  .add_input(*p)
+                  .check_all_same_dtype(false)
+                  .build();
+
+  AT_DISPATCH_ALL_TYPES_AND3(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      at::ScalarType::Bool,
+      iter.dtype(),
+      "bernoulli_xpu",
+      [&] {
+        auto f = BernoulliFunctor<scalar_t>();
+        gpu_kernel(iter, f);
+      });
+}
+
+template <typename scalar_t, typename accscalar_t>
+struct BernoulliScalarFunctor {
+  scalar_t operator()(accscalar_t rand) const {
+    return static_cast<scalar_t>(rand < p_);
+  }
+
+  BernoulliScalarFunctor(accscalar_t p) : p_(p) {}
+
+ private:
+  accscalar_t p_;
+};
+
+template <typename RNG>
+void bernoulli_kernel(TensorIteratorBase& iter, double p, RNG gen) {
+  AT_DISPATCH_ALL_TYPES_AND3(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      at::ScalarType::Bool,
+      iter.dtype(),
+      "bernoulli_scalar_xpu_",
+      [&] {
+        using accscalar_t = at::DiscreteDistributionType<scalar_t>::type;
+        uniform_and_transform<scalar_t, accscalar_t, rand4_engine_calls>(
+            iter, gen, BernoulliScalarFunctor<scalar_t, accscalar_t>(p));
```

Comment:
We need to paths and apply2 to support this?
<img width="952" alt="image" src="https://github.com/intel/torch-xpu-ops/assets/28250760/a63084b5-218b-4a86-b491-d9519ef8a771">